### ISSUE TITLE
bug dans la synthese, la recherche pour une date ne donne pas les relevés avec un horaire 

### ISSUE THREAD

**Issue by RomainBaghi** [PnX-SI/GeoNature#778]
Si on fait une recherche pour une date précise, par exemple le 23/09/2019 (date min = date max dans le formulaire de recherche) on obtient les données de cette date dont l'horaire n'a pas été précisé à la saisie (h = 00:00:00) par contre on loupe toutes les données avec un horaire précisé.

à ce niveau là c'est assez contre intuitif.

Maintenant, s'il on veut afficher ces données avec des horaires définis, on va changer la date max pour le jour suivant. Le problème, c'est qu'on est maintenant parasité par les données du jour suivant...

Solution:  si dans le formulaire on mets date min = date max alors il faut mettre l'heure à 23:59:59 pour date max.

Et cela soulève le fait que les données saisies dans occtax sans horaires définis se voient attribuer l'heure par défaut 00:00:00 pour date min ET date max alors que la logique voudrait que pour date max ce soir 23:59:59 puisque qu'on a pas d'info plus précise.

vous en pensez quoi ?

**Comment by camillemonchicourt** [comment]
Au niveau de la saisie, je pense que quand on ne saisit pas d'heure dans Occtax, il faudrait laisser le champs nul (si cela est possible).

Pour la recherche, si on cherche sur une date, cela devrait remonter toutes les observations de cette date, sans se soucier des heures remplies ou pas.

**Comment by RomainBaghi** [comment]
dans la table t_releves d'occtax il y a des colonnes hour_min, hour_max, qui restes à NULL si il n'a pas eu d'horaires saisis. Donc de ce coté là pas de problème mais quand cela passe à la synthèse, là, on n'a plus que les colonnes date_min et date_max au format timestamp (ex: '1988-10-08 00:00:00'), avec une heure par défaut obligatoire. C'est peut -être à ce niveau là qu'il faut transformer l'heure de date max à 23:59:59 si date_min = date_max ?

**Comment by camillemonchicourt** [comment]
Ah oui c'est vrai, c'est un seul champs timestamp dans la Synthèse. 
Quand on n'a pas renseigné d'heure, c'est un peu faux de mettre 00:00 pour l'heure début et 23:59 pour l'heure de fin.

**Comment by RomainBaghi** [comment]
Eh bien, on fait la même chose pour une obs où on a l'info que sur l'année. On mets bien du 1er janv au 31 décembre. 
C'est là ou le champ qui permettra de définir la précision temporelle sera intéressant (obs ponctuelle ou autre).

**Comment by jbrieuclp** [comment]
Mais du coup on peut aussi se poser la question d'avoir des champs type timestamp (qui intègre l'heure) pour le stockage de la date dans occtax.

**Comment by RomainBaghi** [comment]
oui, c'est pour cela que c'est cohérent d'avoir la colonne hour_min et hour_max dans occtax et effectivement ce serait plus logique de ne pas mettre un timestamp avec heure mais juste une date.
Par contre quand on verse dans la synthèse, il faut se poser la question de ce que l'on fait.

**Comment by gildeluermoz** [comment]
La précision de l'heure dans la synthèse est-elle utile ?
Peut-on considérer que cette information, qui n'est possiblement pas présente dans d'autres sources qu'occtax, ne devrait pas figurer en synthèse ? Dans ce cas, comme le propose Romain, il faudrait passer en champ date simple sur la synthèse.
Une autre solution, plus simple et sans rien modifier dans la base, serait de modifier le comportement de la recherche pour qu'elle se fasse systématiquement entre 00:00 pour date_min et 23:59:59 pour date_max.

**Comment by DonovanMaillard** [comment]
Je partirais plutôt sur la 2eme proposition de Gil. 

Les heures restent à mon avis quelque chose d'informatif sur pas mal de taxons/protocoles : captures chiros, swarmings, durées de pêches électriques, etc. 

L'autre chose à voir c'est que si on supprime à terme l'export d'occtax (#765 ) on rentrerait une information horaire, qui ne serait plus exportable nulle part... 
Que les heures ne soient pas recherchables est en soit pas si grave à mon sens, mais connaitre la durée d'une chasse, d'une capture, d'une pêche, d'un chronoventaire ou autre relevés à terme peut être intéressant je pense.

**Comment by jbdesbas** [comment]
Ca serait effectivement dommage de supprimer l'heure de la synthèse étant donné que celle-ci fait partie du standard INPN.

**Comment by gildeluermoz** [comment]
Oui puisque c'est dans le standard. Effectivement. Par contre, si on mettait aussi dans la synthèse 2 champs hour_min et hour_max et passer le champ date_min et date_max en date au lieu de timestamp ça serait plus simple à gérer sur tous les plans. Je ne me souviens pas pourquoi on a fait d'une façon coté Occtax et d'une autre coté synthèse.
Par contre c'est pas mal de modifs à faire. Que ce soit coté triggers + scripts de migration des bases existantes + éventuellement dans le backend. Fragile tout ça. C'est pour cette raison que j'ai proposé une première approche qui consiste à modifier la recherche.

**Comment by DonovanMaillard** [comment]
Oui, c'est sans doute le plus simple : chercher par défaut de 0:00:00 à 23:59:59 quand rien n'est défini. Et parfaitement cohérent...

**Comment by RomainBaghi** [comment]
Je parlais de remplacer le format timestamp par un format date simple pour les colonnes date_min et date_max dans occtax uniquement. 
Dans la synthèse c'est une autre histoire effectivement.

Donc oui, je pense qu'il faut modifier le formulaire de la synthèse pour que la recherche par date se fasse implicitement de 0:00:00 à 23:59:59, sinon on loupe des obs.

Pour le transfert d'une donnée depuis occtax vers la synthèse, je pensais qu'on pouvait affecter **dans la table synthese** les heures de début et de fin à 0:00:00 et 23:59:59 quand aucun horaire n'a été précisé à la saisie. 
Si l'on ne fait pas cela, on aura une obs qui a été faite du 'XXXX-XX-XX 00:00:00' au 'XXXX-XX-XX 00:00:00' (puisque les colonnes date_min et date_max sont au format timestamp).
Mon raisonnement : Il s'agit d'une donnée qui a été faite ce jour là mais sans précision supplémentaire donc la date de début et la date de fin englobent toute la journée : 'XXXX-XX-XX 00:00:00' et 'XXXX-XX-XX 23:59:59'.

Suite à la remarque de jbdesbas, je comprends que par rapport au standard, l'imprécision temporelle n'est évoquée pour que pour les dates et pas pour les heures. Donc si je comprends bien, on ne transmets pas d'heure (min et max) si l'on a pas l'info. Il faudrait dans ce cas là pouvoir identifier dans la synthèse les obs concernée ? 
Ou plutôt dès que l'on a 00:00:00 pour hmin et hmax, on peut conclure que l'on a pas d'info sur l'heure, et dans ce cas là fini le problème...

**Comment by camillemonchicourt** [comment]
Encore mieux selon moi, si l'on sépare date et heure dans 2 champs, l'idéal est de ne remplir le champs heure que quand une info a été fournie, sinon le laisser vide.

**Comment by TheoLechemia** [comment]
> Encore mieux selon moi, si l'on sépare date et heure dans 2 champs, l'idéal est de ne remplir le champs heure que quand une info a été fournie, sinon le laisser vide.

Oui c'est l'idéal. Comme disais Gil, ça necessite pas mal de transformations, notamment des trigger.
Le fait que l'appli recherche de 00:00 à 23:59 peut faire l'affaire en attendant.

**Comment by TheoLechemia** [comment]
Fait en filtrant sur la date_max <= 23:59:59

**Comment by RomainBaghi** [comment]
super merci !

**Comment by jbdesbas** [comment]
> > Eh bien, on fait la même chose pour une obs où on a l'info que sur l'année. On mets bien du 1er janv au 31 décembre.
> > C'est là ou le champ qui permettra de définir la précision temporelle sera intéressant (obs ponctuelle ou autre).
> 
> Ce n'est pas tout à fait la même chose, les _dateDebut_ et _dateFin_ représentent bien l'imprécision de la date (au sens du SINP). Alors que _heureDebut_ et _heureFin_ représentent une durée d'observation. [cf occtax 3](https://inpn.mnhn.fr/docs/standard/StandardDonneesINPNv3_1.zip)

Pour revenir sur ce propos, je viens de m’apercevoir que le standard [occtax 2](https://inpn.mnhn.fr/docs/standard/Occurrences_de_taxon_v2_0_FINALE_LONGUE.pdf) (le standard d'échange entre plateformes régionales et le niveau national), défini _heureDebut_ et _heureFin_ comme représentant une imprécision OU une durée d'observation :roll_eyes: . Et que donc, par défaut on devrait indiquer 23:59:59 pour l'heure de fin (dans ce dernier cas, l'utilisation de champs dateTime dans GN a du sens).
Il me semble là que c'est surtout au niveau du SINP qu'il y a une incohérence. Comment on gère ça ? On tranche au niveau de GN sur la définition de ces champs, ou on demande à l'INPN de clarifier ça ?

**Comment by camillemonchicourt** [comment]
La version 3 du standard Occurrences de taxons prévoit d'ajout un champs booléen "DateImprecise" au niveau de chaque observation.
Je suis pas trop pour mettre une heure par défaut quand aucune heure n'est renseignée.
